File: apps/ios/Loonext/Features/Contacts/ContactTimelineSection.swift
import SwiftUI

/// The broadcasts that change a customer's history.
///
/// `task.changed` is the real wire name — the DB trigger in
/// `20260702060000_appv2_tasks_attachments_geocode.sql` emits it, and every
/// other client listens for it. An earlier draft here said `task.updated`,
/// which nothing broadcasts, so job rows never revalidated.
private let timelineRefreshEvents: Set<String> = [
    "call.updated",
    "message.created",
    "conversation.updated",
    "task.changed",
]

/// #324 — "what have we done for this customer?", answered by scrolling once.
///
/// D7 threads by recency, so a long relationship is MANY conversations. The
/// prior-conversations list (G6) and the per-contact call history (#205) both
/// already existed and are both still right; what was missing is that they were
/// separate blocks with jobs nowhere, so the question asked before every visit
/// meant opening threads one at a time.
///
/// This sits ABOVE the Calls section rather than replacing it: this is the
/// overview, and Calls stays as the detail view where a voicemail plays in
/// place. Mirror of the Android `ContactTimelineSection`.
///
/// The kinds are told apart by icon and by what the line says, never by being
/// put back into separate lists — merging them is the entire point.
struct ContactTimelineSection: View {
    let graph: AppGraph
    /// Passed in rather than constructed: ContactMutations needs both the api
    /// and the multipart client, and the detail view already holds one.
    let mutations: ContactMutations
    let companyId: String
    let contactId: String
    let onOpenConversation: ((_ conversationId: String) -> Void)?

    @State private var state: LoadState<TimelineLog> = .loading
    @State private var loadingMore = false
    @State private var refreshKey = 0
    /// #517: the roster, so an answered call can say who took it. Best-effort
    /// and out of band — the history must still render if this fails, because
    /// the name is a decoration on a line that already reads correctly.
    @State private var answererNames: [String: String] = [:]

    @Environment(\.appLocale) private var appLocale

    var body: some View {
        ContactSection(
            title: AppStrings.translate(appLocale, "contactsTasks.historySection")
        ) {
            content
        }
        .task(id: "\(contactId)|\(refreshKey)") { await reload() }
        .task(id: companyId) {
            if let page = try? await mutations.members(companyId: companyId) {
                // Named `answererNames` so the property cannot shadow the
                // free `memberNames(_:)` this line calls.
                answererNames = memberNames(page.data, locale: appLocale)
            }
        }
        // The history changes when a text lands, a call ends, or a job moves,
        // so it revalidates on the broadcasts those surfaces already send
        // rather than inventing a fourth event.
        .task(id: contactId) {
            for await event in await graph.realtime.events()
            where timelineRefreshEvents.contains(event.event) {
                refreshKey += 1
            }
        }
        .task(id: contactId) {
            for await _ in await graph.realtime.reconnected() {
                refreshKey += 1
            }
        }
        .resyncOnForeground { refreshKey += 1 }
    }

    @ViewBuilder
    private var content: some View {
        switch state {
        case .loading:
            HStack {
                Spacer()
                ProgressView().controlSize(.small)
                Spacer()
            }
            .padding(.vertical, 12)
        case .failed(let message):
            VStack(alignment: .leading, spacing: 4) {
                Text(message)
                    .font(.golos(12))
                    .foregroundStyle(BrandColor.muted500)
                Button(AppStrings.translate(appLocale, "common.retry")) { refreshKey += 1 }
                    .font(.golos(12, weight: .semibold))
                    .foregroundStyle(BrandColor.olive)
                    .buttonStyle(.plain)
            }
            .padding(.horizontal, 6)
        case .ready(let log):
            let entries = log.entries
            if entries.isEmpty {
                Text(AppStrings.translate(appLocale, "contactsTasks.timelineEmpty"))
                    .font(.golos(12.5))
                    .foregroundStyle(BrandColor.muted500)
                    .frame(maxWidth: .infinity, alignment: .leading)
                    .padding(.horizontal, 6)
            } else {
                readyList(groupTimelineByDay(entries), nextCursor: log.nextCursor)
            }
        }
    }

    private func readyList(
        _ groups: [TimelineDayGroup],
        nextCursor: String?
    ) -> some View {
        VStack(alignment: .leading, spacing: 0) {
            ForEach(groups) { group in
                SectionHeader(label: group.label, count: group.entries.count)
                    .padding(.top, group.id == groups.first?.id ? 4 : 10)
                PaperCard {
                    // Keyed on the dedupe key, not `id`: the three source
                    // tables have independent id spaces, so identity by id
                    // alone could collide across kinds.
                    ForEach(group.entries, id: \.dedupeKey) { entry in
                        TimelineRow(
                            entry: entry,
                            memberNames: answererNames,
                            onOpen: openAction(for: entry)
                        )
                        if entry.dedupeKey != group.entries.last?.dedupeKey {
                            RowDivider().padding(.leading, 42)
                        }
                    }
                }
            }
            if let cursor = nextCursor {
                HStack {
                    Spacer()
                    if loadingMore {
                        ProgressView().controlSize(.small)
                    } else {
                        Button(
                            AppStrings.translate(appLocale, "contactsTasks.showEarlier")
                        ) { loadMore(from: cursor) }
                            .font(.golos(12, weight: .semibold))
                            .foregroundStyle(BrandColor.olive)
                            .buttonStyle(.plain)
                    }
                    Spacer()
                }
                .padding(.top, 8)
            }
        }
    }

    /// Extracted with an explicit type — the same swiftc type-checker guard the
    /// calls section documents. A call that never threaded has nowhere to go,
    /// and a dead tap target is worse than a plain row.
    private func openAction(for entry: TimelineEntry) -> (@MainActor () -> Void)? {
        guard let id = entry.conversation_id, let onOpenConversation else { return nil }
        return { onOpenConversation(id) }
    }

    private func reload() async {
        do {
            let page = try await mutations.timeline(
                companyId: companyId,
                contactId: contactId
            )
            let cached: TimelineLog? = if case .ready(let existing) = state { existing } else { nil }
            // The merge carries the cursor: when it keeps a deeper tail it also
            // keeps the DEEPER cursor, because the fresh page's next_cursor
            // points at the end of page one and adopting it would make "Show
            // earlier" re-request rows already on screen.
            state = .ready(mergeTimelineFirstPage(cached: cached, page: page))
        } catch {
            // A failed background revalidate must NOT throw away a history the
            // reader is looking at — the sibling Calls section guards this the
            // same way. Only a first load with nothing on screen shows an error.
            if case .ready = state { return }
            state = .failed(error.userMessage)
        }
    }

    private func loadMore(from cursor: String) {
        guard !loadingMore else { return }
        loadingMore = true
        Task { @MainActor in
            defer { loadingMore = false }
            do {
                let page = try await mutations.timeline(
                    companyId: companyId,
                    contactId: contactId,
                    cursor: cursor
                )
                guard case .ready(let current) = state else { return }
                state = .ready(appendTimelinePage(current: current, page: page))
            } catch {
                // Keep what is loaded; the button stays, so the retry is one tap.
            }
        }
    }
}

private struct TimelineRow: View {
    let entry: TimelineEntry
    let memberNames: [String: String]
    let onOpen: (@MainActor () -> Void)?

    var body: some View {
        let row = HStack(spacing: 10) {
            Image(systemName: iconName)
                .font(.scaled(15))
                .foregroundStyle(BrandColor.muted500)
                .frame(width: 22)
            VStack(alignment: .leading, spacing: 2) {
                Text(timelineTitle(entry, memberNames: memberNames))
                    .font(.golos(14))
                    .lineLimit(1)
                Text(timelineDetail(entry))
                    .font(.golos(12))
                    .foregroundStyle(BrandColor.muted500)
                    .lineLimit(1)
            }
            Spacer(minLength: 8)
            Text(timeLabel)
                .font(.golos(11))
                .foregroundStyle(BrandColor.muted500)
        }
        .padding(.horizontal, 12)
        .padding(.vertical, 10)
        .contentShape(Rectangle())

        if let onOpen {
            Button(action: onOpen) { row }.buttonStyle(.plain)
        } else {
            row
        }
    }

    private var iconName: String {
        switch entry.kind {
        case "call": return "phone"
        case "task": return entry.done == true ? "checkmark.circle" : "circle"
        default: return "bubble.left"
        }
    }

    private var timeLabel: String {
        // parseWireTimestamp and timelineShortDate: the app's own helpers. The
        // `.formatted(.dateTime...)` API is used nowhere else here, and Swift
        // only compiles in CI, so an unproven API is a 25-minute round trip.
        guard let date = parseWireTimestamp(entry.occurred_at) else { return "" }
        return timelineShortDate(date, calendar: .current, format: "h:mm a")
    }
}

File: apps/android/app/src/main/kotlin/com/loonext/android/features/contacts/ContactTimelineSection.kt
package com.loonext.android.features.contacts

import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.outlined.CheckCircle
import androidx.compose.material.icons.outlined.Call
import androidx.compose.material.icons.outlined.ChatBubbleOutline
import androidx.compose.material.icons.outlined.RadioButtonUnchecked
import androidx.compose.material3.Icon
import androidx.compose.material3.LoadingIndicator
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.key
import androidx.compose.runtime.mutableIntStateOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.loonext.android.core.data.CacheKeys
import com.loonext.android.core.i18n.LocalAppLocale
import com.loonext.android.core.i18n.t
import com.loonext.android.AppGraph
import com.loonext.android.ui.common.LoadState
import com.loonext.android.ui.common.PaperCard
import com.loonext.android.ui.common.RowDivider
import com.loonext.android.ui.common.SectionHeader
import com.loonext.android.ui.common.SkeletonListRow
import com.loonext.android.features.thread.memberNames
import com.loonext.android.ui.common.rememberCacheFirst
import com.loonext.android.ui.common.rememberHaptics
import java.time.Instant
import java.time.ZoneId
import java.time.format.DateTimeFormatter
import kotlinx.coroutines.launch

/**
 * #324 — "what have we done for this customer?", answered by scrolling once.
 *
 * D7 threads by recency, so a long relationship is MANY conversations. The
 * prior-conversations list (G6) and the per-contact call history (#205) both
 * already existed and are both still right; what was missing is that they were
 * separate blocks with jobs nowhere, so the question asked before every visit
 * meant opening threads one at a time.
 *
 * This sits ABOVE the Calls section rather than replacing it: this is the
 * overview, and Calls stays as the detail view where a voicemail plays in
 * place.
 *
 * The kinds are told apart by icon and by what the line says, never by being
 * put back into separate lists — merging them is the entire point.
 */
@Composable
internal fun ContactTimelineSection(
    graph: AppGraph,
    mutations: ContactMutations,
    companyId: String,
    contactId: String,
    onOpenConversation: ((conversationId: String) -> Unit)?,
    modifier: Modifier = Modifier,
) {
    var refreshKey by remember(contactId) { mutableIntStateOf(0) }
    var loadingMore by remember(contactId) { mutableStateOf(false) }
    val scope = rememberCoroutineScope()
    val haptics = rememberHaptics()
    // #228: the day headings are built inside a LazyListScope, which is not a
    // composition, so the reader's language is read here and carried in.
    val locale = LocalAppLocale.current

    // #176 cache-first, like the Calls section: a reopened contact paints its
    // history instantly while the first page revalidates, and the merge keeps
    // any deeper pages the user already loaded.
    val cacheKey = CacheKeys.contactTimeline(companyId, contactId)
    val state = rememberCacheFirst(
        cache = graph.storeCache,
        key = cacheKey,
        refreshKey = refreshKey,
    ) {
        mergeTimelineFirstPage(
            graph.storeCache.flowOf<ContactTimelineLog>(cacheKey).value,
            mutations.timeline(companyId, contactId),
        )
    }

    // The history changes when a text lands, a call ends, or a job moves — so
    // it revalidates on the same broadcasts those surfaces already listen to
    // rather than inventing a fourth event.
    LaunchedEffect(contactId) {
        graph.realtime.events.collect { event ->
            when (event.event) {
                // `task.changed` is the wire name the tasks trigger emits
                // (20260702060000). An earlier draft said `task.updated`,
                // which nothing broadcasts, so job rows never revalidated.
                "call.updated", "message.created", "conversation.updated", "task.changed" ->
                    refreshKey++
            }
        }
    }
    LaunchedEffect(contactId) {
        graph.realtime.reconnected.collect { refreshKey++ }
    }

    // #517: the roster, so an answered call can say who took it. Best-effort
    // and out of band — the history must still render if this fails, because
    // the name is a decoration on a line that already reads correctly.
    var memberNames by remember(companyId) { mutableStateOf(emptyMap<String, String>()) }
    LaunchedEffect(companyId) {
        runCatching { mutations.members(companyId) }
            .getOrNull()
            ?.let { page -> memberNames = memberNames(page.data) }
    }

    ContactSection(t("contactsTasks.historySection"), modifier) {
        when (val current = state) {
            is LoadState.Loading -> PaperCard(Modifier.fillMaxWidth()) {
                SkeletonListRow(avatar = false)
                RowDivider()
                SkeletonListRow(avatar = false)
            }

            is LoadState.Failed -> Column(Modifier.padding(start = 6.dp)) {
                Text(
                    current.message,
                    style = MaterialTheme.typography.bodySmall.copy(fontSize = 12.sp),
                    color = MaterialTheme.colorScheme.onSurfaceVariant,
                )
                TextButton(
                    onClick = {
                        haptics.tap()
                        refreshKey++
                    },
                    contentPadding = PaddingValues(0.dp),
                ) { Text(t("common.retry")) }
            }

            is LoadState.Ready -> {
                val entries = current.value.entries
                if (entries.isEmpty()) {
                    Text(
                        t("contactsTasks.timelineEmpty"),
                        style = MaterialTheme.typography.bodySmall.copy(fontSize = 12.5.sp),
                        color = MaterialTheme.colorScheme.onSurfaceVariant,
                        modifier = Modifier.padding(start = 6.dp),
                    )
                } else {
                    val groups = remember(entries) { groupTimelineByDay(entries) }
                    groups.forEachIndexed { groupIndex, (day, rows) ->
                        SectionHeader(
                            timelineDayLabel(day, locale = locale),
                            Modifier.padding(top = if (groupIndex == 0) 0.dp else 10.dp),
                            count = rows.size,
                        )
                        PaperCard(Modifier.fillMaxWidth()) {
                            rows.forEachIndexed { index, entry ->
                                key("${entry.kind}:${entry.id}") {
                                    TimelineRow(
                                        entry = entry,
                                        memberNames = memberNames,
                                        onOpen = entry.conversation_id
                                            ?.takeIf { onOpenConversation != null }
                                            ?.let { id -> { onOpenConversation?.invoke(id) } },
                                    )
                                    if (index < rows.lastIndex) RowDivider()
                                }
                            }
                        }
                    }
                    if (current.value.nextCursor != null) {
                        Box(
                            Modifier
                                .fillMaxWidth()
                                .padding(top = 4.dp),
                            contentAlignment = Alignment.Center,
                        ) {
                            if (loadingMore) {
                                LoadingIndicator(Modifier.size(28.dp))
                            } else {
                                TextButton(onClick = {
                                    val cursor = current.value.nextCursor
                                        ?: return@TextButton
                                    haptics.tap()
                                    loadingMore = true
                                    scope.launch {
                                        try {
                                            val page = mutations.timeline(
                                                companyId, contactId, cursor = cursor,
                                            )
                                            // Append onto whatever the cache holds
                                            // NOW: a silent revalidate may have
                                            // landed since the tap.
                                            val base = graph.storeCache
                                                .flowOf<ContactTimelineLog>(cacheKey).value
                                                ?: current.value
                                            graph.storeCache.put(
                                                cacheKey,
                                                appendTimelinePage(base, page),
                                            )
                                        } catch (_: Exception) {
                                            // Keep what is loaded; the button stays,
                                            // so the retry is one tap.
                                        } finally {
                                            loadingMore = false
                                        }
                                    }
                                }) { Text(t("contactsTasks.showEarlier")) }
                            }
                        }
                    }
                }
            }
        }
    }
}

private val TIME_LABEL: DateTimeFormatter = DateTimeFormatter.ofPattern("h:mm a")

@Composable
private fun TimelineRow(
    entry: TimelineEntry,
    memberNames: Map<String, String>,
    onOpen: (() -> Unit)?,
) {
    val haptics = rememberHaptics()
    val locale = LocalAppLocale.current
    Row(
        Modifier
            .fillMaxWidth()
            .then(
                // A call that never threaded has nowhere to go, and a dead tap
                // target is worse than a plain row.
                if (onOpen != null) {
                    Modifier.clickable {
                        haptics.tap()
                        onOpen()
                    }
                } else {
                    Modifier
                },
            )
            .padding(horizontal = 12.dp, vertical = 10.dp),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(10.dp),
    ) {
        Icon(
            imageVector = when {
                entry.kind == "call" -> Icons.Outlined.Call
                entry.kind == "task" && entry.done == true -> Icons.Outlined.CheckCircle
                entry.kind == "task" -> Icons.Outlined.RadioButtonUnchecked
                else -> Icons.Outlined.ChatBubbleOutline
            },
            contentDescription = null,
            modifier = Modifier.size(20.dp),
            tint = MaterialTheme.colorScheme.onSurfaceVariant,
        )
        Column(Modifier.weight(1f)) {
            Text(
                timelineTitle(entry, memberNames, locale),
                style = MaterialTheme.typography.bodyMedium,
                maxLines = 1,
            )
            Text(
                timelineDetail(entry, locale),
                style = MaterialTheme.typography.bodySmall.copy(fontSize = 12.sp),
                color = MaterialTheme.colorScheme.onSurfaceVariant,
                maxLines = 1,
            )
        }
        Text(
            TIME_LABEL.format(
                Instant.parse(entry.occurred_at).atZone(ZoneId.systemDefault()),
            ),
            style = MaterialTheme.typography.labelSmall,
            color = MaterialTheme.colorScheme.onSurfaceVariant,
        )
    }
}
